spark/src/test/scala/org/apache/spark/sql/delta/FindLastCompleteCheckpointSuite.scala

```diff
@@ -0,0 +1,293 @@
+    versions.flatMap { version =>
+      FileNames.checkpointFileWithParts(logPath, version, numParts).map(pathToFileStatus(_, length))
+    }
+  }
+
+  private def checksumFiles(logPath: Path, versions: Seq[Long]): Seq[FileStatus] = {
+    versions.map { version => pathToFileStatus(FileNames.checksumFile(logPath, version)) }
+  }
+
+  test("findLastCompleteCheckpoint without any argument") {
+    withTempDir { dir =>
+      val log = DeltaLog.forTable(spark, dir.getAbsolutePath)
+      val logPath = log.logPath
+      val logStore = log.store.asInstanceOf[CustomListingLogStore]
+
+      // Case-1: Multiple checkpoint exists in table dir
+      logStore.customListingResult = Some(
+        commitFiles(logPath, 0L to 3000) ++
+          singleCheckpointFiles(logPath, Seq(100, 200, 1000, 2000))
+      )
+      assert(log.findLastCompleteCheckpointBefore().contains(CheckpointInstance(version = 2000)))
+      assert(logStore.listFromCount == 1)
+      assert(logStore.elementsConsumedFromListFromIter == 3005)
+      logStore.reset()
+
+      // Case-2: No checkpoint exists in table dir
+      logStore.customListingResult = Some(commitFiles(logPath, 0L to 3000))
+      assert(log.findLastCompleteCheckpointBefore().isEmpty)
+      assert(logStore.listFromCount == 1)
+      assert(logStore.elementsConsumedFromListFromIter == 3001)
+      logStore.reset()
+
+      // Case-3: Multiple checkpoints for same version exists in table dir
+      logStore.customListingResult = Some(
+        commitFiles(logPath, 0L to 3000) ++
+          singleCheckpointFiles(logPath, Seq(100, 200, 1000, 2000)) ++
+          multipartCheckpointFiles(logPath, Seq(300, 2000), numParts = 4)
+      )
+      assert(log.findLastCompleteCheckpointBefore().contains(
+        CheckpointInstance(version = 2000, Format.WITH_PARTS, numParts = Some(4))))
+      assert(logStore.listFromCount == 1)
+      assert(logStore.elementsConsumedFromListFromIter == 3013)
+      logStore.reset()
+    }
+  }
+
+  test("findLastCompleteCheckpoint with an upperBound which exists") {
+    withTempDir { dir =>
+      val log = DeltaLog.forTable(spark, dir.getAbsolutePath)
+      val logPath = log.logPath
+      val logStore = log.store.asInstanceOf[CustomListingLogStore]
+      logStore.reset()
+
+      // Case-1: The upperBound exists and it should not be returned
+      logStore.customListingResult = Some(
+        commitFiles(logPath, 0L to 3000) ++
+          singleCheckpointFiles(logPath, Seq(100, 200, 1000, 2000))
+      )
+      assert(
+        log.findLastCompleteCheckpointBefore(Some(CheckpointInstance(version = 2000)))
+          .contains(CheckpointInstance(version = 1000)))
+      assert(logStore.listFromCount == 1)
+      assert(logStore.elementsConsumedFromListFromIter == 1002 + 2) // commits + checkpoint
+      logStore.reset()
+
+      // Case-2: The exact upperBound (a multi-part checkpoint) doesn't exist but another single
+      // part checkpoint for same version exists.
```

Comment:
No. The single-part checkpoint comes before multi-part checkpoint and the below assertion verifies that it is returned.

      assert(log.findLastCompleteCheckpointBefore(Some(sentinelCheckpoint))
        .contains(CheckpointInstance(version = 2000)))

Reply:
ah, I had that backward sorry -- we _prefer_ the multi-part checkpoint because we prefer the latest checkpoint and multi-part compares "later" than single-part. So if we ask for a checkpoint _before_ the multi-part checkpoint, then we would indeed expect to find the single-part.